### PR TITLE
Make Time update_with_instant pub

### PR DESCRIPTION
**Pull request by ostwilkens**
# Objective

A lot of systems depend on Time to implement frame-independent logic. 
To be able to write tests for these systems, one needs to be able to mock Time. By mocking time, it's possible to write tests that confirm systems are frame-independent. 

## Solution

To mock Time, one needs to be able to manually update the Time resource with an Instant defined by the developer. 
There is a function for this, but it's currently only public within the crate. I propose this function is exposed to the developer, to make it easy to write tests for systems utilizing the Time resource.

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
bors try

**Comment by alice-i-cecile** [comment]
Welcome! I agree with the need for this, and think this is an excellent, simple solution.

As updating the time step manually is typically going to be a mistake in end user code outside of tests, we should add careful doc strings to both this method and the Time type to explain the correct usage :)

**Comment by alice-i-cecile** [comment]
Actually, this may also be useful for running simulations at faster than real time too. I've seen that requested before too; we should try to document that use case too. A full example might be better for that (you'll need to mess with one of the default plugins), but that work could be saved for a later PR.

**Comment by LegNeato** [comment]
Perhaps only make this pub under #[test] to remove that footgun?

**Comment by bjorn3** [comment]
> Perhaps only make this pub under #[test] to remove that footgun?

That is not possible. `#[cfg(test)]` only works when the current crate is being tested, not any dependent crate.

**Comment by mockersf** [comment]
> Actually, this may also be useful for running simulations at faster than real time too. I've seen that requested before too; we should try to document that use case too. A full example might be better for that (you'll need to mess with one of the default plugins), but that work could be saved for a later PR.

It's probably a bad idea to manually update with an instant in this case, as it would break all deltas exposed by `Time`. To change the speed from real time it's much simpler to apply a factor to everything you get from `Time`.

To make this useful outside of test, one would need to disable the `time_system` system
https://github.com/bevyengine/bevy/blob/f71f93c843ea03d38477353ac02e3854053e27f9/crates/bevy_core/src/time/time.rs#L87-L89

I don't think that's a good idea, so I would prefer clear doc comments explaining this is only useful for tests. Could a feature to enable test helpers work?

**Comment by alice-i-cecile** [comment]
>To make this useful outside of test, one would need to disable the time_system system

This was what I had in mind; it's much more useful in scientific settings but I could see it being useful for accelerated simulation in ordinary games (e.g. for machine learning or automated integration tests).

Out of scope for this though, and I'd want to try testing that strategy on a full game to see if anything breaks before recommending it.

**Comment by ostwilkens** [comment]
In the case of non-linear or scaled time as a feature in games, it might be recommended to keep track of simulated time as a separate resource. In such a case it would be nice to be able to wrap or alias Time, and modify it as needed for your usecase. It's something that should be made possible by this change. 

I'll add the doc strings discussed.

**Comment by ostwilkens** [comment]
I'm not sure if it's feasible, or a good idea, but one way to prevent breaking the default Time resource would be to somehow declare the built-in resource as read-only (outside the bevy crate).

**Comment by ostwilkens** [comment]
Sorry about that; rebased from 0.5.0 release.. should I open a new PR?

**Comment by mockersf** [comment]
> Sorry about that; rebased from 0.5.0 release.. should I open a new PR?

Yup, if you can't clean this PR, it would probably be best to open a new one

**Comment by ostwilkens** [comment]
Sorry about the fuzz, I think it's cleaned up up now. Open to input on the doc string and test.

**Comment by mdickopp** [comment]
This change would be very useful to me as well, so I would like to help. What needs to be done to get this ready to be merged?

File: crates/bevy_core/src/time/time.rs [hunk 1]
```diff
@@ -31,7 +31,8 @@ impl Time {
         self.update_with_instant(now);
     }
 
-    pub(crate) fn update_with_instant(&mut self, instant: Instant) {
+    /// Update time with a specified [`Instant`], for use in tests. Should never be called on the default Time resource.
```
Comment (alice-i-cecile) [inline crates/bevy_core/src/time/time.rs:34]:
```suggestion
    /// Update time with a specified [`Instant`]
    ///
    /// This method is for use in tests, and should never be called on the default Time resource.
```

**Comment by alice-i-cecile** [comment]
So to get this merged, the next step is that we need at least two approvals, including from community members, and all blocking concerns addressed.

On my end, I want to see:

1. Better doc strings, explaining exactly what the risks are when calling this inside an `App`.
2. Sign-off from @maniwani, who's been redesigning how Bevy handles time.

**Comment by maniwani** [comment]
I'm cool with making `update_with_instant` `pub` since `update` already is.

I will say tho that the featured test seems a little misleading.

At the end it highlights the results between 30 FPS and 60 FPS being similar after a single frame as if that would be the case forever (which is false). Integrating by `time.delta_seconds()` does decouple motion from the framerate but not in a way that is repeatable or consistent between different FPS.

You only get real framerate independence if you have a constant step size (and then `FixedTimestep` or a dedicated thread).

> something something running simulations faster than real-time

Is what #3002 is for.

File: crates/bevy_core/src/time/time.rs (hunk 1 above)
Comment (maniwani) [reply crates/bevy_core/src/time/time.rs:34]: Just inaccurate timekeeping in general. I would keep the recommendation to *only* use this to write tests.

**Review by maniwani — APPROVED** [review]
I'm cool with making `update_with_instant` `pub` since `update` already is.

I will say tho that the featured test seems a little misleading.

At the end it highlights the results between 30 FPS and 60 FPS being similar after a single frame as if that would be the case forever (which is false). Integrating by time.delta_seconds() does decouple motion from the framerate but not in a way that is repeatable or consistent between different FPS.

You only get real framerate independence if you have a constant step size (and then `FixedTimestep` or a dedicated thread).

> something something running simulations faster than real-time

Is what #3002 is for.

File: crates/bevy_core/src/time/time.rs (hunk 1 above)
Comment (maniwani) [reply crates/bevy_core/src/time/time.rs:34]: Just inaccurate timekeeping in general. I would keep the recommendation to *only* use this to write tests.

**Comment by devil-ira** [comment]
I believe most people understand framerate-independence to mean that motion is separate from the framerate. i.e. the game does not appear to move in slow-motion when the framerate dips, or vise versa. Rather than separating simulation-time from real-time in a fixed timestep setting.

With that in mind, I don't think the test is pretending to be more accurate than it is.

File: tests/how_to_test_systems.rs [hunk 2]
```diff
@@ -95,3 +111,60 @@ fn spawn_enemy_using_input_resource() {
...
+        // Remove mocked Time
+        world.remove_resource::<Time>();
+
+        // Return resulting position
+        *world.get_entity(entity).unwrap().get::<Position>().unwrap()
+    };
+
+    // Test at 30 and 60 fps
+    let result_30fps = test_fps(30);
+    let result_60fps = test_fps(60);
```
Comment (devil-ira) [inline tests/how_to_test_systems.rs:163]:
```suggestion
    let result_15fps = test_fps(15);
    let result_30fps = test_fps(30);
    let result_60fps = test_fps(60);
    let result_120fps = test_fps(120);
    let result_240fps = test_fps(240);
```
Perhaps we should test some more extreme framerates as well.

**Comment by maniwani** [comment]
> I believe most people understand framerate-independence to mean [...]

In my experience, time-related things confuse users a lot. 

I understand this is the kind of test OP wanted to mock, but I prefer don't feature a test _with this premise_ on the off-chance someone interprets it as an "officially-endorsed" way of handling physics.

This is all fine for UI, etc, but if users want stable game physics, they should use a fixed timestep (with interpolation to make it look smooth) because that's the standard and correct way to do it.

> With that in mind, I don't think the test is pretending to be more accurate than it is.

I was more pointing out that the test seems really arbitrary if the point is just to confirm that someone didn't forget to multiply by `time.delta_seconds()` somewhere—it checks that the motion didn't desync by more than an arbitrary amount under arbitrary FPS values after an arbitrarily small duration.

**Comment by devil-ira** [comment]
> seems pretty convoluted if the point is just to confirm that we didn't forget to multiply by time.delta_seconds().
Sorry, I probably got too preoccupied with your comment. I agree with you here. what's being tested is basically how much float addition drifts over 30 iterations compared to 60 iterations. I'm in favor of removing the test.

**Comment by mdickopp** [comment]
I'm not the OP, so I cannot update the OP's branch. Should I open a new PR?

> I'm in favor of removing the test.

I would like to rework the test into an example demonstrating how the function can be used in a test. What do you think?

**Comment by alice-i-cecile** [comment]
> I'm not the OP, so I cannot update the OP's branch. Should I open a new PR?

Yep sounds great: just link this PR in the description and we can close this one out.

> I would like to rework the test into an example demonstrating how the function can be used in a test. What do you think?

Yes please! I think a doc test on `Time` itself would be a nice place for this :) That feels like the most likely place users will look if they're wondering about "how do I mock this".

**Comment by ostwilkens** [comment]
I don't currently have the environment set up to assist on this, I'm glad it's being looked into. Feel free to rework it in a new PR.

**Comment by mdickopp** [comment]
I opened #4469 as a follow-up/replacement PR to this one.

**Comment by alice-i-cecile** [comment]
Closed in favor of #4469

**Pull request closed without merging**